tls/s2n_connection.c

```diff
@@ -147,30 +147,28 @@ struct s2n_connection *s2n_connection_new(s2n_mode mode)
     conn = (struct s2n_connection *)(void *)blob.data;
 
     if (s2n_is_in_fips_mode()) {
-        conn->config = &s2n_default_fips_config;
+        s2n_connection_set_config(conn, s2n_fetch_default_fips_config());
     } else {
-        conn->config = &s2n_default_config;
+        s2n_connection_set_config(conn, s2n_fetch_default_config());
     }
 
     if (mode == S2N_CLIENT) {
-        /* At present s2n is not suitable for use in client mode, as it
-         * does not perform any certificate validation. However it is useful
-         * to use S2N in client mode for testing purposes. An environment
-         * variable is required to be set for the client mode to work.
-         */
+        /* we need to do more testing on our x.509 code. Until then use with caution */
         if (getenv("S2N_ENABLE_CLIENT_MODE") == NULL) {
             GUARD_PTR(s2n_free(&blob));
             S2N_ERROR_PTR(S2N_ERR_CLIENT_MODE_DISABLED);
         }
-        /* S2N does not have it's own x509 Certificate parser, so Client Mode should only be used for testing purposes.
-         * In Client mode, we skip Cert Validation and assume that all Server RSA x509 Certs are valid. */
-        conn->config = &s2n_unsafe_client_testing_config;
+
+        /* For the moment, this still needs to be modified via s2n_connection_set_config() after this function is called. To use the secure configuration */
+        s2n_connection_set_config(conn, s2n_fetch_unsafe_client_testing_config());
     }
 
     conn->mode = mode;
     conn->blinding = S2N_BUILT_IN_BLINDING;
     conn->close_notify_queued = 0;
     conn->session_id_len = 0;
+    conn->verify_host_fn = NULL;
```

Comment:
Shouldn't be since you couldn't call s2n_connection_set_config() prior to s2n_connection_new()

But the connection callback, will always take precedence over the config callback. This is to allow you to set a config wide callback for convenience, but most of the time (at least in my case), I want to share the config, while having a separate hostname verification for each connection.

In particular, in our SDKs, we want to be able to have customizable hostname verification for handling S3 wildcard certificates better. For example, we can't do host addressing for some buckets due to too many '.' characters in the hostname. This allows our SDKs to override that behavior when they need to.